File: String/Balanced Parentheses/balanced_parantheses.py
# A small check for balanced parantheses in Python 3
# Added regex as a little extra

import re

any_parantheses = re.compile("[\[({})\]]+")
opening_parantheses = ['[', '(', '{']
closing_parantheses = [']', ')', '}']


def is_balanced(expr):
    stack = []
    n = len(expr)
    for i in range(0, n):
        if (expr[i] in opening_parantheses):
            stack.append(expr[i])
        elif (expr[i] in closing_parantheses):
            matching_paranthesis = opening_parantheses[closing_parantheses.
                                                       index(expr[i])]
            if (stack.pop() != matching_paranthesis):
                return False

    return True if len(stack) == 0 else False


def main():
    expr = input(
        "Please enter an expression with parantheses ('[','(','{','}',')',']') or 'test' for a fixed test: "
    )
    if (expr == "test"):
        expr = "({}[](([{()}])))"

    if (not any_parantheses.match(expr)):
        print("No parantheses found!")
    elif (is_balanced(expr)):
        print("%s is balanced!" % expr)
    else:
        print("%s is not balanced!" % expr)


if __name__ == '__main__':
    main()

File: String/Balanced Parentheses/JavaScript/balanced_parantheses.js
/* Balanced Parantheses */

function isBalanced(expr) {
  var s = [];
  for (var i = 0; i < expr.length; i++) {
    var ch = expr.charAt(i);
    if (ch == '(' || ch == '{' || ch == '[')
      s.push(ch);
    else if (ch == ')' || ch == '}' || ch == ']') {
      if (ch == ')' && top(s) == '(')
        s.pop();
      else if (ch == '}' && top(s) == '{')
        s.pop();
      else if (ch == ']' && top(s) == '[')
        s.pop();
      else
        return false;
    }
  }

  return s.length == 0;
}

// Finding the Top of the Stack s.
// For empty stack Top is -1
function top(s) {
  var len = s.length
  return len > 0 ? s[len - 1] : -1;
}

var expressions = [
  "{[()]}",
  "{[(])}",
  "{{[[(())]]}}",
  "]{[()]()}",
  "{[()])(}"
];
expressions.forEach(function(e) {
  var areBalanced = isBalanced(e);
  if (areBalanced)
    console.log(e + " is Balanced");
  else
    console.log(e + " is Not Balanced");
});